modules/catalog-compaction/README.md

```diff
@@ -1 +1,101 @@
+
+1. **Replicas update**. Updates all replication groups with determined minimum begin
+   time among all active RW transactions in the cluster. After some time (see below for details)
+   these timestamps are published and become available for the next phase.
+
+2. **Compaction**. Using the timestamps published on the previous stage calculates
+   the minimum required version of the catalog and perform compaction.
+
+Publishing timestamps can take a long time, and the success of compaction depends on more
+than just these timestamps. That's why both stages run in parallel. Thus, the compaction
+stage uses the result of the replicas update calculated at one of the previous iterations.
+To minimize the number of network requests, both processes run simultaneously and use a common
+[request](src/main/java/org/apache/ignite/internal/catalog/compaction/message/CatalogCompactionMinimumTimesRequest.java)
+to collect timestamps from the entire cluster in one round trip.
+
+### Replicas update stage
+
+![Replicas update](tech-notes/replicas-update.png)
+
+This stage consists of the following steps:
+
+1. Each node uses [ActiveLocalTxMinimumBeginTimeProvider](../transactions/src/main/java/org/apache/ignite/internal/tx/ActiveLocalTxMinimumBeginTimeProvider.java)
+   to determine the minimum begin time among all local active RW transactions and sends it to coordinator.
+2. Coordinator calculates global minimum and sends it to all nodes using [CatalogCompactionPrepareUpdateTxBeginTimeMessage](src/main/java/org/apache/ignite/internal/catalog/compaction/message/CatalogCompactionPrepareUpdateTxBeginTimeMessage.java).
+3. Each node stores this time within replication groups for which the local node is the leader  
+   (using [UpdateMinimumActiveTxBeginTimeReplicaRequest](../partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/network/replication/UpdateMinimumActiveTxBeginTimeReplicaRequest.java)
+   and [UpdateMinimumActiveTxBeginTimeCommand](../partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/network/command/UpdateMinimumActiveTxBeginTimeCommand.java). 
+4. This timestamp (let's call it `minTxTime`) is published (become available to compaction process) only
+   after checkpoint (flushing partition data to disk).
+
+### Compaction stage
+
+![Replicas update](tech-notes/compaction.png)
+
+1. Each node determines the local minimum required time, this consists of the following steps:
+   1. Using introduced `RaftGroupStateProvider` determine minimum time among all published
+      timestamps (`minTxTime`) in local replication groups.
+   2. If `minTxTime` is not published yet, the current iteration of compaction is aborted.
+   3. Select minimum between determined minimum `minTxTime` and current `low watermark`.
+2. Each node sends to the coordinator the calculated local minimum timestamp,
+   as well as a set of local replication groups that was used to calculate the local minimum.
```

Comment:
as well as to the set of local replication groups that was used to calculate the local minimum.

Reply:
Didn't get it. what "as well as TO the set" mean here.
We sending not only timestamp but also a set

Reply:
Then i misunderstood the statement, my bad.